llvm/lib/Transforms/IPO/SampleProfileMatcher.cpp

```diff
@@ -782,6 +782,26 @@ bool SampleProfileMatcher::functionMatchesProfileHelper(
   float Similarity = 0.0;
 
   const auto *FSFlattened = getFlattenedSamplesFor(ProfFunc);
+  // Check if the function is top-level function. For extended profile format,
+  // if a function profile is unused and it's top-level, even if the profile is
+  // matched, it's not found in the profile. This is because sample reader only
+  // read the used profile at the beginning, we need to read the profile
+  // on-demand. Also save it into the FlattenedProfiles for future look-up.
+  if (!FSFlattened) {
+    DenseSet<StringRef> TopLevelFunc;
+    TopLevelFunc.insert(ProfFunc.stringRef());
+    SampleProfileMap TopLevelProfile;
+    Reader.readOnDemand(TopLevelFunc, TopLevelProfile);
```

Comment:
How often does this happen on this path (coming from LCS I think)? And how useful it is to load profile here? 

Asking because we probably don't want to load many functions just to do the similarity checks during LCS, as loading profile is kind of expensive?

Reply:
How about we add a switch to on/off this feature? Just in case this (loading profile for matched functions) overall becomes expensive.